documentation/design_docs/design/brokered-addressspace.adoc

```diff
@@ -0,0 +1,106 @@
+== why
+
+The standard address space is designed for scalable messaging using the dispatch router to allow
+high throughput and many connections. This comes at the expense of some broker features such as
+message grouping, transactions and other broker-specific features.
+
+The brokered address space is designed to support broker-specific features, at the cost of limited
+scale in terms of the number of connections and addresses.
+
+== what/how
+
+The brokered address space will not change any APIs in the standard address space, but will involve
+changes to some components in order to share code and reduce runtime overhead. The brokered address
+space will consist of a 'broker' and an 'agent' process.
+
+=== address model
+
+The address model will be extendended with the 'brokered' address space and the address types. To begin
+with, the supported address types will be 'queue' and 'topic'. Address-type specific features will
+be added as properties.
+
+=== service broker
+
+The service broker cannot expose address types 'queue' multiple times, so it needs to have a mapping
+for address types in each address space type.
+
+=== address controller
+
+The address controller needs to generalize the logic needed to watch standard address space configs to
+handle brokered address space configs. It will not contain any address-space specific logic for the
+brokered address space, and the parts that are standard address space-specific, will (eventually,
+not part of the first iteration) moved.
+
+=== console/generic container
+
+The console will be refactored to support the brokered address space and generalized into a generic
+container runtime called 'enmasse agent' that can run multiple plugins. The initial set of plugins will be 'console',
+and 'brokered-controller'. The address configuration will be retrieved using an API that hides the
+kubernetes interaction.
+
+=== console plugin
+
+The console plugin will implement the console as we have today in the standard address space, but
+modified to support the brokered address space.
+
+=== brokered-controller plugin
+
+This plugin will be responsible for creating/deleting queues and topics on brokers. It monitors the
+address configured within its namespace, and ensures that the brokers state matches that of the
+namespace. The plugin will also be responsible for updating the address resource status based on the
+state of that address within the broker.
+
+There are two alternatives for connecting the controller to the broker. Either
+
+1. through a broker service that maps to the broker cluster
+2. through a connector in the broker that connects to the controller
+
+Alternative 1. gives us load balancing over a broker cluster for free, and since we will need it for
+clients connecting to the service anyway, it does not cost any extra. Alternative 2. requires the
```

Comment:
We may want a distinct service for internal use(?).

Reply:
Yes, I think we should do that from the start.